### ISSUE TITLE
Feature Request - ValidatePath option

### ISSUE THREAD

**Issue by jschpp** [PowerShell/PowerShell#10249]
# Summary of the new feature/enhancement
There should be an easy option to validate path parameters.

> As a developer I want to validate a path parameter with a keyword to stop using the same validation pattern over and over again.

At the moment I need to use this pattern all the time:
```powershell
param (
    [ValidateNotNullOrEmpty()]
    [ValidateScript({
        if(-not (Test-Path $_)){
            throw
        }
    })]
    [string]
    $Path
)
```

I would like to replace it with something like this:
```powershell
param (
    [ValidatePath()]
    [string]
    $Path
)
```

There was already a similarly issue #5647 which was bigger in scope than this.

# Proposed technical implementation details (optional)

I'v tried my hand at creating this validator #10250

**Comment by KirkMunro** [comment]
Are you going to handle any PowerShell provider? Or just the `FileSystem` provider? `ValidatePath` as a parameter validator should handle any provider, not just files.

I find myself wondering if this should be a validator. Internally if it requires invoking `Test-Path` again and again, that could be expensive when used at scale, and it's quite specific as far as validators go.

Since it seems from your example, and from the PR the way it is written right now, that your interest is in file/directory validation, I wonder if it would be better to have:

1. A pair of transformation attributes, `[FileInfo()]` and `[DirectoryInfo()]`, that can convert a string into a corresponding `FileInfo` or `DirectoryInfo` object (these objects support string conversion already, but transformation attributes are necessary if we want these to support PSPath syntax as well).

1. A property validator, `[ValidateProperty('PropertyName','Value')]` that would validate that a particular object has a property with a particular value.

The property validator alone would be generally much more useful than `ValidatePath`, and would allow for something like this:

```PowerShell
function Use-Directory {
    [CmdletBinding()]
    param(
        [Parameter(Position=0, Mandatory, ValueFromPipelineByPropertyName)]
        [ValidateProperty('Exists',$true)]
        [System.IO.DirectoryInfo]
        [Alias('FullName')]
        $Directory
    )
}
Get-Item C:\ | Use-Directory
```

If you wanted PSPath support as well, you could use the proposed `[DirectoryInfo()]` transformation attribute, and write it like this instead:

```PowerShell
function Use-Directory {
    [CmdletBinding()]
    param(
        [Parameter(Position=0, Mandatory, ValueFromPipelineByPropertyName)]
        [ValidateProperty('Exists',$true)]
        [System.IO.DirectoryInfo]
        [DirectoryInfo()]
        [Alias('PSPath')]
        $Directory
    )
}
Get-Item C:\ | Use-Directory
```

**Comment by iSazonov** [comment]
> As a developer I want to validate a path parameter with a keyword to stop using the same validation pattern over and over again.

The motivation is not clear because the pattern is as simple as Test-Path.

**Comment by vexx32** [comment]
Correct. However, use of `[ValidateScript()]` is needlessly syntax-heavy and awkward. PowerShell is an administration tool for computer systems; built-in validation for paths should be provided out of the box, in my opinion. Additionally, having parameter validation for paths run on pre-compiled code is probably likely to give some performance benefits in parameter binding for what seems to be very little cost. 🙂

**Comment by iSazonov** [comment]
>having parameter validation for paths run on pre-compiled code is probably likely to give some performance benefits in parameter binding for what seems to be very little cost. 

The attribute will only slow down. If we want to be fast we should move from string paths to objects with pre-processed paths (like `FileInfo`)

**Comment by vexx32** [comment]
More than ValidateScript? Really? 

That kind of conversion could be done by the validation attribute anyway, couldn't it?

**Comment by msftrncs** [comment]
Potential problem with this idea.  If you send an entire pipeline full of paths, doesn't the validation have to succeed on all paths before command can commence?  In many cases that would not be desired to have all the paths checked before work starts, and that just one bad path prevents any of the process.  For single path arguments/pipelines its no big deal.   However, I think the thought process was that you could detect on your own if the path is bad, and display a better error message or take other action, better than just a simple validation.

Additionally, if the paths are validated at the binding of a large/long/slow script, the paths may be invalid by the time the script performs the work, so the script always needs to be prepared for dealing with the exception of a missing path anyway.

Validation of other kinds of inputs makes sense, because they are not moving targets.

So in short, you shouldn't even be using ValidateScript to validate a moving target such as a file path at binding time.  Instead you should be prepared to deal with file path issues using try/catch/finally or trap.

I could see wanting to validate if a path contains any invalid components or syntax, since on some OS's the syntax of a file path is different.

But these are just opinions...

**Comment by vexx32** [comment]
@msftrncs indeed, but the same issue is faced with ValidateScript and Test-Path; it is up to the command implementor to determine if the path is critical or non-critical (whether it should terminate the whole pipeline as a parameter-binding issue).

In most cases where this would be used, the path argument would not be a pipeline parameter and typically critical to the command function. For example, a Write-Log command where it takes data / strings as pipeline input and writes to a single path.

**Comment by KirkMunro** [comment]
@vexx32 The key difference between `ValidateScript` and a `ValidatePath` attribute that internally invokes `Test-Path` is visibility of the performance hit. I think built-in validators should be very performant and not invoke PowerShell to do their validation.

Also would `ValidatePath` return true if the item behind the path exists? Or if the path was valid? What about validation against a specific provider, since many functions are written to only work with a limited set (often just one) of providers? What about wildcard validation? There are many validation scenarios, and a single `ValidatePath` validator does not unambiguously define which one it validates.

**Comment by vexx32** [comment]
Pretty simple to resolve all your concerns there, Kirk. Plenty of validators allow parameters, we can simply leave it up to the command implementer to decide exactly what they're after.

```powershell
function Write-Thing {
    param(
        [Parameter(Mandatory)]
        [ValidatePath(<# MustExist #> $true, <# ProviderName #> 'FileSystem')]
        [string]
        $Path,

        [Parameter(Mandatory, ValueFromPipeline)]
        [object]
        $InputObject
    )
    process {
        Set-Content -Value $InputObject -Path $Path
    }
}
